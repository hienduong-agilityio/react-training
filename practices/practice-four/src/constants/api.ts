export const HTTP_METHODS = {
  GET: 'GET',
  POST: 'POST',
  PUT: 'PUT',
  DELETE: 'DELETE'
};

export const ROOT_API = import.meta.env.VITE_API_ENDPOINT;

export const API_URL = {
  PROJECT: `${ROOT_API}/projects`
};
