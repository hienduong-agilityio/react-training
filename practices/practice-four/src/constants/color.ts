// Consider using enums instead of constants
export const COLORS_CLASS = {
  DEFAULT: 'color-default',
  PRIMARY: 'color-primary',
  SUCCESS: 'color-success',
  WARNING: 'color-warning',
  DANGER: 'color-danger '
};
