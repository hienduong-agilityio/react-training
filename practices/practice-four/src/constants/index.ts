// Index constants files
export * from './api';
export * from './color';
export * from './filerSearchOption';
export * from './route';
export * from './rowPerPage';
export * from './store';
export * from './validationMessages';
