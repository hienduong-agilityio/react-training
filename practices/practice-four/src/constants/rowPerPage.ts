export const ROWS_PER_PAGE_OPTIONS = [
  { value: 5, label: '5' },
  { value: 10, label: '10' },
  { value: 15, label: '15' },
  { value: 20, label: '20' },
  { value: 30, label: '30' },
  { value: 40, label: '40' },
  { value: 50, label: '50' }
];
