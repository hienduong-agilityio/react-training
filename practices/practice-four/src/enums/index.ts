// Index Enum files
export * from './status';
export * from './theme';
