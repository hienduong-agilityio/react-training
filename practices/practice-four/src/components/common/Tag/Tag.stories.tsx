// Types
import type { Meta, StoryObj } from '@storybook/react';

// Component
import Tag from './index';

// Enum
import { COLORS } from '@/enums';

const meta = {
  title: 'Components/common/Tag',
  component: Tag,
  parameters: {
    layout: 'centered'
  },
  tags: ['autodocs'],
  argTypes: {
    color: {
      description:
        'Color of the Tag. Can be one of the COLORS enums: "default", "primary", "success", "warning", "danger".',
      control: { type: 'select', options: Object.values(COLORS) }
    },
    customClasses: {
      description: 'Custom CSS classes to be applied to the Tag.',
      control: { type: 'text' }
    },
    children: {
      description: 'Content inside the Tag.',
      control: { type: 'text' }
    }
  }
} satisfies Meta<typeof Tag>;

export default meta;

type Story = StoryObj<typeof meta>;

export const TagDefault: Story = {
  args: {
    children: 'default',
    color: COLORS.DEFAULT
  },
  parameters: {
    docs: {
      description: {
        story: 'Default Tag with the "default" color scheme. Suitable for displaying neutral or secondary information.'
      }
    }
  }
};

export const AllColors: Story = {
  render: (args) => (
    <div style={{ display: 'flex', gap: '8px' }}>
      <Tag {...args} children='Default' color={COLORS.DEFAULT} />
      <Tag {...args} children='Primary' color={COLORS.PRIMARY} />
      <Tag {...args} children='Success' color={COLORS.SUCCESS} />
      <Tag {...args} children='Warning' color={COLORS.WARNING} />
      <Tag {...args} children='Danger' color={COLORS.DANGER} />
    </div>
  ),
  args: {
    children: '',
    color: COLORS.DEFAULT
  },
  parameters: {
    docs: {
      description: {
        story:
          'Displays all available colors for the Tag component. Demonstrates different color variants: default, primary, success, warning, and danger.'
      }
    }
  }
};
