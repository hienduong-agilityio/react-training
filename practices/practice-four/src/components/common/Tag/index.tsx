// Libraries
import { memo } from 'react';
import classNames from 'classnames';

// Enum
import { COLORS } from '@/enums';

// Helpers
import { getColorClasses } from '@/helpers';

export interface ITagProps {
  // children: The children of the Tag.
  children: React.ReactNode;
  // color: Color of the Tag. Can be one of COLORS enums.
  color?: COLORS;
  // customClasses: Custom variant class for the Tag.
  customClasses?: string;
}

const defaultClasses: string =
  'inline-flex gap-[6px] rounded font-medium leading-[18px] tracking-[0.36px] flex items-center justify-center';

/**
 * Tag component
 *
 * @returns {JSX.Element} - Tag element.
 */
const Tag = ({ children, color = COLORS.DEFAULT, customClasses = '' }: ITagProps): JSX.Element => {
  const colorClasses = getColorClasses(color);

  const tagClasses: string = classNames(customClasses, defaultClasses, colorClasses);

  return <span className={tagClasses}>{children}</span>;
};

export default memo(Tag);
