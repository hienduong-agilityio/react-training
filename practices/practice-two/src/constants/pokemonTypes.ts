import { CHIP_COLOR } from '@components/common/Chip';

const POKEMON_CHECKBOX_TYPES: {
  type: CHIP_COLOR;
  checked: boolean;
}[] = [
  { type: CHIP_COLOR.FIRE, checked: false },
  { type: CHIP_COLOR.BUG, checked: false },
  { type: CHIP_COLOR.WATER, checked: false },
  { type: CHIP_COLOR.FLYING, checked: false },
  { type: CHIP_COLOR.POISON, checked: false },
  { type: CHIP_COLOR.NORMAL, checked: false },
  { type: CHIP_COLOR.ROCK, checked: false },
  { type: CHIP_COLOR.GROUND, checked: false },
  { type: CHIP_COLOR.FIGHTING, checked: false },
  { type: CHIP_COLOR.GHOST, checked: false },
  { type: CHIP_COLOR.PSYCHIC, checked: false },
  { type: CHIP_COLOR.ICE, checked: false },
  { type: CHIP_COLOR.DRAGON, checked: false },
  { type: CHIP_COLOR.DARK, checked: false },
  { type: CHIP_COLOR.FAIRY, checked: false },
  { type: CHIP_COLOR.ELECTRIC, checked: false },
  { type: CHIP_COLOR.STEEL, checked: false },
  { type: CHIP_COLOR.GRASS, checked: false }
];

export { POKEMON_CHECKBOX_TYPES };
