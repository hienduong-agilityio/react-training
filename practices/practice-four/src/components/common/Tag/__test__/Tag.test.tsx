// Libraries
import { render, screen } from '@testing-library/react';

// Components
import Tag, { ITagProps } from '@/components/common/Tag';

// Enums
import { COLORS } from '@/enums';

// Constants
import { COLORS_CLASS } from '@/constants';

/*
Tag Component
  √ should match the snapshot
  √ should render the children content
  √ should apply custom classes
 */
describe('Tag Component', () => {
  let renderComponent: (props?: Partial<ITagProps>) => ReturnType<typeof render>;

  beforeEach(() => {
    renderComponent = (props: Partial<ITagProps> = {}) => {
      return render(<Tag {...props}>Tag Content</Tag>);
    };
  });

  it('should match the snapshot', () => {
    const { container } = renderComponent({ color: COLORS.SUCCESS });
    expect(container).toMatchSnapshot();
  });

  it('should render the children content', () => {
    renderComponent();

    const tagElement = screen.getByText('Tag Content');
    expect(tagElement).toBeInTheDocument();
    expect(tagElement).toHaveClass(COLORS_CLASS.DEFAULT);
  });

  it('should apply custom classes', () => {
    const customClass = 'my-custom-class';
    renderComponent({ customClasses: customClass });

    const tagElement = screen.getByText('Tag Content');
    expect(tagElement).toHaveClass(customClass);
  });
});
