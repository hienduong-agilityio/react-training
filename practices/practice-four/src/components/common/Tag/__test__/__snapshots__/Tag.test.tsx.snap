// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tag Component should match the snapshot 1`] = `
<div>
  <span
    class="inline-flex gap-[6px] rounded font-medium leading-[18px] tracking-[0.36px] flex items-center justify-center color-success"
  >
    Tag Content
  </span>
</div>
`;
